### PR TITLE
DataFixers [1.15x] JavaDoc time

### PR DESCRIPTION
**Pull request by i509VCB**
This is meant to be a replacement to #371 due to 1.15 conflicts being so great. All that needs to be done is some wiki writing and addition of the rest of ModDataVersion entiries being added into NBT tags where nessecary.

### CODE REVIEW

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/TypeReferenceHelper.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ * These types only apply the singular transformation, so any custom elements will have to be added manually via {@link DSL#fields(String, TypeTemplate)} or {@link DSL#optionalFields(String, TypeTemplate)}
+ */
+public interface TypeReferenceHelper {
+	/**
+	 * Gets the instance of the {@link TypeReferenceHelper}.
+	 */
+	TypeReferenceHelper HELPER = TypeRefHelperImpl.INSTANCE;
+
+	/**
+	 * Registers a type with a inTile element.
```
Comment (liach) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/TypeReferenceHelper.java:41]: Hmm, what is in tile?

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerImpl.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+		@Override
+		public Message newMessage(String message) {
+			return new SimpleMessage("[Fabric-DataFixer] ");
+		}
+
+		@Override
+		public Message newMessage(String message, Object... params) {
+			return new SimpleMessage("[Fabric-DataFixer] " + params);
+		}
+	});
```
Comment (liach) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerImpl.java:56]: uh what?

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerInitalizerClient.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+package net.fabricmc.fabric.impl.datafixer;
+
+import net.fabricmc.api.ClientModInitializer;
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+import net.fabricmc.fabric.api.event.client.ClientTickCallback;
+
+@Environment(EnvType.CLIENT)
+public final class FabricDataFixerInitalizerClient implements ClientModInitializer {
```
Comment (liach) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerInitalizerClient.java]: `FabricClientDataFixerInitalizer`

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerInitalizerCommon.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package net.fabricmc.fabric.impl.datafixer;
+
+import net.fabricmc.api.ModInitializer;
+import net.fabricmc.fabric.api.event.server.ServerStartCallback;
+import net.fabricmc.fabric.impl.datafixer.test.TestObjects;
+import net.fabricmc.fabric.impl.datafixer.test.TestUtil;
+
+public final class FabricDataFixerInitalizerCommon implements ModInitializer {
```
Comment (liach) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerInitalizerCommon.java]: also put common in the middle of the word

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/fixes/BiomeRenameFix.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+	@Override
+	protected TypeRewriteRule makeRule() {
+		// First we must get the type represented by the biome. We do this and get a NamedType in return
+		Type<Pair<String, String>> type_1 = DSL.named(TypeReferences.BIOME.typeName(), DSL.namespacedString());
+
+		// Next we must make sure the input type matches the biome or else data corruption could occur.
+		if (!Objects.equals(type_1, this.getInputSchema().getType(TypeReferences.BIOME))) {
+			throw new IllegalStateException("Biome type is not what was expected.");
+		} else {
```
Comment (liach) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/fixes/BiomeRenameFix.java]: no need for this `else`

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/fixes/BlockEntityRenameFix.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	@SuppressWarnings("unchecked")
+	@Override
+	public TypeRewriteRule makeRule() {
+		TaggedChoiceType<String> originalTypeChoice = (TaggedChoiceType<String>) this.getInputSchema().findChoiceType(TypeReferences.BLOCK_ENTITY);
+		TaggedChoiceType<String> newTypeChoice = (TaggedChoiceType<String>) this.getOutputSchema().findChoiceType(TypeReferences.BLOCK_ENTITY);
+		Type<Pair<String, String>> blockEntityType = DSL.named(TypeReferences.ENTITY_NAME.typeName(), DSL.namespacedString()); // Why on earth does this work with ENTITY_NAME.
+
+		if (!Objects.equals(this.getOutputSchema().getType(TypeReferences.ENTITY_NAME), blockEntityType)) { // I mean seriously I didn't change that field but it works. Then again ENTITY TypeReference does not refer to ENTITY_NAME in any instances I've seen in registerTypes.
+			throw new IllegalStateException("BlockEntity name type is not what was expected.");
+		} else {
```
Comment (liach) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/fixes/BlockEntityRenameFix.java]: no need for this `else`

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/fixes/BlockEntityRenameFix.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+			return TypeRewriteRule.seq(this.fixTypeEverywhere(this.fixName, originalTypeChoice, newTypeChoice, (ops) -> (pair) -> pair.mapFirst((originalName) -> {
+				String possiblyNamedString = this.rename(originalName);
+				Type<?> originalType = originalTypeChoice.types().get(originalName);
+				Type<?> newType = newTypeChoice.types().get(possiblyNamedString);
+
+				if (!newType.equals(originalType, true, true)) {
+					throw new IllegalStateException(String.format("Dynamic type check failed: %s not equal to %s", newType, originalType));
+				} else {
```
Comment (liach) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/fixes/BlockEntityRenameFix.java]: no need for this `else`

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/mixin/MixinPersistentState.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+package net.fabricmc.fabric.impl.datafixer.mixin;
+
+import org.spongepowered.asm.mixin.Mixin;
+
+import net.minecraft.world.PersistentState;
+
+@Mixin(PersistentState.class)
+public class MixinPersistentState {
+	// Add our mod's data versions to method_17919 (INJECT @ after putInt)
```
Comment (liach) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/mixin/MixinPersistentState.java:25]: I fear this may blow up the size of these saved data files.

**Review by liach — COMMENTED** [review]
Looks fairly good.

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/mixin/MixinPersistentState.java (hunk 8 above)
Comment (i509VCB) [reply fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/mixin/MixinPersistentState.java:25]: Well based on the subclasses I see of PersistentState, the only one I would have a worry about having to implement this into would be ScoreboardState since objectives are a TypeReference the game can fix.

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/TypeReferenceHelper.java (hunk 1 above)
Comment (i509VCB) [reply fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/TypeReferenceHelper.java:41]: inTile had something to do with the arrow entities and tridents, maybe it's the state of the block the object is within or something similar?

Comment (liach) [reply fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/TypeReferenceHelper.java:41]: mind describe this more clearly in the jd?

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerImpl.java (hunk 2 above)
Comment (i509VCB) [reply fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerImpl.java:56]: Well that was a brainfart, I threw the wrong parameter into the simple message.

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/TypeReferenceHelper.java (hunk 1 above)
Comment (Juuxel) [reply fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/api/datafixer/v1/TypeReferenceHelper.java:41]: Is this like the `inBlockState` in the projectile NBT?
nbt format on mcwiki

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/test/TestUtil.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+		@Override
+		public CompoundTag toTag(CompoundTag tag) {
+			CompoundTag tempTag = super.toTag(tag);
+			tempTag.putInt("ground_beef", beef);
+			return tempTag;
+		}
+	}
+
+	public static class TestBEBlock extends Block implements BlockEntityProvider {
+		public TestBEBlock(Settings block$Settings_1) {
```
Comment (modmuss50) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/test/TestUtil.java:74]: Didnt realise that was a valid field name.

Might want to change it anyway

**Comment by Player3324** [comment]
I think this API should be carefully evaluated to support lazy initialization, MP client or saves created with the same versions will never need to initialize the data fixer infrastructure. It uses an entry point, so it should be already mostly there, but this needs to be explicitly tested.

Mod data may have a different versioning requirement than vanilla, how should that be handled? Does data fixer support multiple independently versioned parts or is it limited to its one global version number?

**Comment by i509VCB** [comment]
Mod versioning for data versions is left to the mod devs to handle. However I would say the convention should be starting from 0. The setting of version flags in world data is handled if you've registered a Datafixer.

For the case of lazy init, mojang doesn't even do that, and the update method in the Datafixer will do nothing if the data version is equal.

On the first point, I'll need a bit of time to writeup sonething about it.

**Comment by i509VCB** [comment]
@sfPlayer1 so I've thought over a bit, MC does not lazy init the Datafixer at all because of the dynamic nature of chunk loading. Take the fact chunks could go years without being loaded, infact in some worlds a chunk might have not been touched since 1.2.5. Of course based on schema 501 adding polar bears I can assume the furthest back this system would go is to about 1.9.

The only way to possibly have a lazy init would be a Datafixer mode which would have to fix the entire world which on some servers make take hours. Since a Datafixer mode would add a bunch of unneeded complexity I would personally avoid it.

Also Datafixers are created before the client and server start since the constructors obviously take a DataFixer.

**Comment by modmuss50** [comment]
If im not mistaken, vanilla already has a way to fully convert the world.

**Comment by i509VCB** [comment]
I think that would be --forceUpgrade

I know for sure most people would be too lazy to have to type that when updating, most would probably want drop and go

**Comment by i509VCB** [comment]
So I feel this PR has been slow on my part because it is a slow process to document the shear scale of the API. I feel I need to approach this in a more incremental manner. So currently I believe I will approach like this:

PR 1: Allow creation and building of datafixers

PR 2 (or maybe part of the first tbf): Simple data fixes, renaming blocks, items and biomes.

PR 3: TypeReferences for modded (Block)Entities.

PR 4: Complex data fixes, entity rename, block entity rename, block entity data transformation, entity data transformation.

@liach what do you think of this as a roadmap?

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerImpl.java [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+
+import net.fabricmc.fabric.api.datafixer.v1.DataFixerEntrypoint;
+import net.fabricmc.fabric.api.datafixer.v1.DataFixerHelper;
+import net.fabricmc.loader.api.FabricLoader;
+
+public final class FabricDataFixerImpl implements DataFixerHelper {
+	public static final Logger LOGGER = LogManager.getLogger("Fabric-DataFixer", new AbstractMessageFactory() {
+		@Override
+		public Message newMessage(String message, Object... params) {
+			return new SimpleMessage("[Fabric-DataFixer] " + message);
```
Comment (liach) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerImpl.java:54]: Why is this needed?

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerImpl.java [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+	 * Represents Minecraft's Built in DataFixer.
+	 */
+	public static final class VanillaDataFixers {
+		private static final int LATEST_VANILLA_SCHEMA_VERSION = DataFixUtils.makeKey(SharedConstants.getGameVersion().getWorldVersion());
+
+		static {
+			LOGGER.info("Started with Vanilla-DataFixer version: " + LATEST_VANILLA_SCHEMA_VERSION);
+		}
+
+		public static final BiFunction<Integer, Schema, Schema> VANILLA_DATAFIXER = (version, parent) -> Schemas.getFixer().getSchema(LATEST_VANILLA_SCHEMA_VERSION);
```
Comment (liach) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/FabricDataFixerImpl.java:215]: should put the lambda body in a static method declaration than let the compiler generate a method (see how java does `Objects#equals`)

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/mixin/MixinNbtHelper.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+import net.minecraft.datafixers.DataFixTypes;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.NbtHelper;
+
+import net.fabricmc.fabric.impl.datafixer.FabricDataFixerImpl;
+
+@Mixin(NbtHelper.class)
+public class MixinNbtHelper {
+	@Inject(at = @At("RETURN"), method = "update(Lcom/mojang/datafixers/DataFixer;Lnet/minecraft/datafixers/DataFixTypes;Lnet/minecraft/nbt/CompoundTag;II)Lnet/minecraft/nbt/CompoundTag;", cancellable = true)
+	private static void updateModFixers(DataFixer vanillaDataFixer, DataFixTypes dataFixTypes, CompoundTag inputTag$unusued, int vanillaDynamicDataVersion, int vanillaRuntimeDataVersion, CallbackInfoReturnable<CompoundTag> cir) {
```
Comment (liach) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/mixin/MixinNbtHelper.java:34]: `inputTag$unusued` typo

File: fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/mixin/Schema100Accessor.java [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+import com.mojang.datafixers.types.templates.TypeTemplate;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.gen.Invoker;
+
+import net.minecraft.datafixers.schemas.Schema100;
+
+@Mixin(Schema100.class)
+public interface Schema100Accessor {
+	// registerTypeWithArmorAndToolSlots
+	@Invoker("method_5195")
```
Comment (liach) [inline fabric-datafixer-v1/src/main/java/net/fabricmc/fabric/impl/datafixer/mixin/Schema100Accessor.java:16]: is this method still unmapped?

**Comment by i509VCB** [comment]
Superseded by #522.
